File: backend/app/services/ml_service.py
import os
import json
import asyncio
import logging
from typing import Dict, List, Optional, Any
from pathlib import Path

# Try to import ML dependencies, fall back gracefully if not available
try:
    import numpy as np
    import pandas as pd
    import joblib
    import torch
    from transformers import AutoTokenizer, AutoModelForSequenceClassification
    from sentence_transformers import SentenceTransformer, util
    ML_AVAILABLE = True
except ImportError as e:
    logging.warning(f"ML dependencies not available: {e}. Using fallback logic.")
    ML_AVAILABLE = False
    # Create mock objects for graceful fallback
    np = None
    pd = None
    joblib = None
    torch = None

from app.models.types import (
    MLCampaignOptimizationRequest, MLCampaignOptimizationResponse,
    MLCreativeScoreRequest, MLCreativeScoreResponse
)

logger = logging.getLogger(__name__)

# Global variables for models
campaign_model = None
feature_columns = None
nlp_models = {}
distilbert_model = None
distilbert_tokenizer = None

# Label mapping for DistilBERT (from README.txt)
DISTILBERT_LABEL_MAPPING = {0: 3, 1: 4, 2: 5, 3: 6, 4: 7, 5: 8}

async def load_ml_models():
    """Load all ML models"""
    global campaign_model, feature_columns, nlp_models, distilbert_model, distilbert_tokenizer

    if not ML_AVAILABLE:
        logger.info("ML dependencies not available, skipping model loading")
        return

    logger.info("Starting ML model loading...")

    # Load campaign optimization model
    try:
        models_path = Path("models")
        model_file = models_path / "campaign_optimizer_usd.pkl"
        features_file = models_path / "model_feature_columns_usd.json"

        if model_file.exists() and features_file.exists():
            campaign_model = joblib.load(str(model_file))
            with open(features_file, "r") as f:
                feature_columns = json.load(f)
            logger.info("✅ Campaign optimization model loaded successfully")
        else:
            logger.warning(f"❌ Campaign model files not found at {models_path}")
            logger.info(f"Expected files: {model_file}, {features_file}")
    except Exception as e:
        logger.error(f"❌ Error loading campaign model: {e}")

    # Load DistilBERT creative scoring model
    try:
        distilbert_path = Path("models/distilbert_creative_scorer")
        if distilbert_path.exists():
            distilbert_tokenizer = AutoTokenizer.from_pretrained(str(distilbert_path))
            distilbert_model = AutoModelForSequenceClassification.from_pretrained(str(distilbert_path))
            distilbert_model.eval()
            logger.info("✅ DistilBERT creative scoring model loaded successfully")
        else:
            logger.warning(f"❌ DistilBERT model not found at {distilbert_path}")
    except Exception as e:
        logger.error(f"❌ Error loading DistilBERT model: {e}")

    # Load NLP models for semantic analysis
    try:
        # Sentence embedder for semantic similarity
        nlp_models['embedder'] = SentenceTransformer("all-MiniLM-L6-v2")
        logger.info("✅ Sentence embedder loaded successfully")

        # Skip paraphraser for now to speed up loading
        logger.info("⏭️ Skipping paraphraser model to speed up loading")

    except Exception as e:
        logger.error(f"❌ Error loading NLP models: {e}")

    # Log final status
    models_loaded = []
    if campaign_model is not None:
        models_loaded.append("Campaign Optimizer")
    if distilbert_model is not None:
        models_loaded.append("DistilBERT Creative Scorer")
    if nlp_models.get('embedder') is not None:
        models_loaded.append("Sentence Embedder")

    logger.info(f"🎉 ML models loaded: {', '.join(models_loaded) if models_loaded else 'None'}")

class MLService:

    @staticmethod
    def build_features_row(total_budget: float, split_budgets: Dict[str, float],
                          aov: float, creative_quality: float, campaign_days: int,
                          target_margin: float, age: int, gender: str, income_level: str):
        """Build feature row for campaign model prediction"""
        channels = ["instagram", "google", "tiktok", "facebook", "youtube", "linkedin"]

        pct_vals = {f"pct_{ch}": (split_budgets[ch]/total_budget if total_budget > 0 else 0.0)
                    for ch in channels}

        feat = {
            **{f"budget_{ch}": split_budgets[ch] for ch in channels},
            **pct_vals,
            "total_budget": float(total_budget),
            "aov": float(aov),
            "creative_quality": float(creative_quality),
            "campaign_days": int(campaign_days),
            "target_margin": float(target_margin),
            "age": int(age),
            "gender": gender,
            "income_level": income_level
        }

        df = pd.DataFrame([feat])
        df = pd.get_dummies(df, columns=["gender", "income_level"])

        # Ensure all columns are present
        for col in feature_columns:
            if col not in df.columns:
                df[col] = 0

        return df[feature_columns]

    @staticmethod
    def generate_candidates(total_budget: float, channels: List[str], K: int = 500, seed: int = 42):
        """Generate candidate budget allocations"""
        if not ML_AVAILABLE or np is None:
            # Fallback: generate simple variations
            import random
            random.seed(seed)
            candidates = []
            for _ in range(min(K, 50)):  # Limit fallback candidates
                # Generate random splits
                splits = [random.random() for _ in channels]
                total = sum(splits)
                normalized = [s/total for s in splits]

                budgets = {ch: round(total_budget * frac, 2) for ch, frac in zip(channels, normalized)}
                # Adjust for rounding errors
                diff = round(total_budget - sum(budgets.values()), 2)
                if abs(diff) >= 0.01:
                    largest = max(budgets, key=budgets.get)
                    budgets[largest] = round(budgets[largest] + diff, 2)
                candidates.append(budgets)
            return candidates

        np.random.seed(seed)
        samples = np.random.dirichlet(np.ones(len(channels)), size=K)
        candidates = []

        for samp in samples:
            budgets = {ch: round(float(total_budget * frac), 2) for ch, frac in zip(channels, samp)}
            # Adjust for rounding errors
            diff = round(total_budget - sum(budgets.values()), 2)
            if abs(diff) >= 0.01:
                largest = max(budgets, key=budgets.get)
                budgets[largest] = round(budgets[largest] + diff, 2)
            candidates.append(budgets)

        return candidates

    @staticmethod
    async def optimize_campaign_budget(request: MLCampaignOptimizationRequest) -> MLCampaignOptimizationResponse:
        """Optimize campaign budget allocation"""
        global campaign_model, feature_columns

        # If ML is not available or model not loaded, use fallback logic
        if not ML_AVAILABLE or not campaign_model or not feature_columns:
            return await MLService._optimize_campaign_budget_fallback(request)

        try:
            channels = ["instagram", "google", "tiktok", "facebook", "youtube", "linkedin"]
            candidates = MLService.generate_candidates(request.total_budget, channels, K=500)

            # Score all candidates
            scored = []
            for cand in candidates:
                try:
                    feat_row = MLService.build_features_row(
                        request.total_budget, cand, request.aov, request.creative_quality,
                        request.campaign_days, request.target_margin, request.age,
                        request.gender, request.income_level
                    )
                    pred_rev = float(campaign_model.predict(feat_row)[0])
                    pred_roi = (pred_rev - request.total_budget) / request.total_budget if request.total_budget > 0 else -9999

                    scored.append({
                        "split": cand,
                        "pred_revenue": pred_rev,
                        "pred_roi": pred_roi
                    })
                except Exception as e:
                    logger.warning(f"Error scoring candidate: {e}")
                    continue

            if not scored:
                return await MLService._optimize_campaign_budget_fallback(request)

            # Get best candidate
            best = max(scored, key=lambda x: x["pred_roi"])

            # Calculate confidence based on variance in top predictions
            top_5_rois = sorted([s["pred_roi"] for s in scored], reverse=True)[:5]
            confidence = max(0.5, 1.0 - (np.std(top_5_rois) * 2))

            warning = None
            if best["pred_roi"] < 0:
                warning = "⚠️ Model predicts this campaign may be unprofitable under given inputs."

            return MLCampaignOptimizationResponse(
                recommended_split=best["split"],
                predicted_revenue=round(best["pred_revenue"], 2),
                predicted_roi=max(0.0, round(best["pred_roi"], 4)),
                confidence_score=round(confidence, 2),
                warning=warning
            )
        except Exception as e:
            logger.error(f"Error in campaign optimization: {e}")
            return await MLService._optimize_campaign_budget_fallback(request)

    @staticmethod
    async def score_creative_content(request: MLCreativeScoreRequest) -> MLCreativeScoreResponse:
        """Score creative content using DistilBERT and NLP models"""
        global distilbert_model, distilbert_tokenizer

        # If DistilBERT model is not available, use fallback
        if not ML_AVAILABLE or not distilbert_model or not distilbert_tokenizer:
            return await MLService._score_creative_content_fallback(request)

        try:
            # Combine title and description for DistilBERT scoring
            combined_text = f"{request.title}. {request.description}. {request.cta}"

            # Get DistilBERT prediction
            inputs = distilbert_tokenizer(combined_text, truncation=True, padding=True, return_tensors="pt")
            with torch.no_grad():
                outputs = distilbert_model(**inputs)
                probs = torch.softmax(outputs.logits, dim=-1).tolist()[0]

            # Convert probabilities to score (3-8 scale)
            expected_score = sum(DISTILBERT_LABEL_MAPPING[i] * prob for i, prob in enumerate(probs))
            distilbert_score = min(10, max(1, expected_score * 1.25))  # Scale to 1-10

            # Enhance with semantic analysis if available
            semantic_boost = 0
            if nlp_models.get('embedder'):
                try:
                    # Reference phrases for good marketing copy
                    good_marketing_phrases = [
                        "limited time offer", "exclusive deal", "act now", "save money",
                        "premium quality", "satisfaction guaranteed", "free shipping",
                        "best value", "top rated", "customer favorite"
                    ]

                    embeddings = nlp_models['embedder'].encode([combined_text] + good_marketing_phrases, convert_to_tensor=True)
                    similarities = util.cos_sim(embeddings[0], embeddings[1:]).cpu().numpy()
                    semantic_boost = float(np.max(similarities)) * 2  # Boost up to 2 points
                except Exception as e:
                    logger.warning(f"Error in semantic analysis: {e}")

            # Final score calculation
            final_score = min(10, max(1, distilbert_score + semantic_boost))

            # Component scores (more granular breakdown)
            title_score = min(10, max(1, len(request.title.split()) * 1.2 + semantic_boost))
            desc_score = min(10, max(1, len(request.description.split()) * 0.6 + distilbert_score * 0.3))
            cta_score = 8.0 if any(word in request.cta.lower() for word in ["buy", "shop", "get", "try", "now"]) else 6.0
            channel_fit = distilbert_score * 0.8  # DistilBERT considers overall quality

            # Generate feedback based on scores
            feedback = []
            improvements = {}

            if title_score < 7:
                feedback.append("Title could be more engaging - try adding urgency or emotional triggers")
                improvements["title"] = MLService.generate_title_improvements(request.title, request.channel)

            if desc_score < 7:
                feedback.append("Description needs stronger call-to-action or more compelling benefits")
                improvements["description"] = MLService.generate_description_improvements(request.description, request.channel)

            if cta_score < 7:
                feedback.append("Call-to-action could be more action-oriented and specific")
                improvements["cta"] = MLService.generate_cta_improvements(request.cta, request.channel)

            if final_score >= 8:
                feedback.append("🎉 Excellent creative! This should perform very well.")
            elif final_score >= 6:
                feedback.append("Good creative with room for improvement.")
            else:
                feedback.append("Creative needs significant improvements for better performance.")

            return MLCreativeScoreResponse(
                channel=request.channel,
                scores={
                    "title": round(title_score, 1),
                    "description": round(desc_score, 1),
                    "cta": round(cta_score, 1),
                    "channel_fit": round(channel_fit, 1),
                    "final": round(final_score, 1)
                },
                feedback=feedback,
                improvements=improvements
            )

        except Exception as e:
            logger.error(f"Error in DistilBERT creative scoring: {e}")
            return await MLService._score_creative_content_fallback(request)

    @staticmethod
    def generate_title_improvements(title: str, channel: str) -> List[str]:
        """Generate title improvement suggestions"""
        improvements = []

        if channel.lower() == "google":
            improvements.extend([
                f"{title} | 50% Off Today",
                f"Best {title} Deals 2024",
                f"Official {title} - Free Shipping"
            ])
        elif channel.lower() in ["instagram", "tiktok"]:
            improvements.extend([
                f"🔥 {title} - Limited Time!",
                f"{title} 😍 Don't Miss Out!",
                f"VIRAL: {title} Everyone Loves ✨"
            ])
        elif channel.lower() == "linkedin":
            improvements.extend([
                f"{title} - Professional Solutions",
                f"Industry-Leading {title} for Business",
                f"Boost ROI with {title}"
            ])
        else:
            improvements.extend([
                f"{title} - Shop Now & Save Big",
                f"Premium {title} Collection",
                f"Limited Stock: {title} Available"
            ])

        return improvements[:3]

    @staticmethod
    def generate_description_improvements(description: str, channel: str) -> List[str]:
        """Generate description improvement suggestions"""
        improvements = []
        base_desc = description[:100] + "..." if len(description) > 100 else description

        if channel.lower() == "google":
            improvements.extend([
                f"{base_desc} Order now with fast, free shipping and 30-day guarantee!",
                f"{base_desc} Save big today with our exclusive deals - limited time only!",
                f"{base_desc} Join thousands of satisfied customers - shop now!"
            ])
        elif channel.lower() in ["instagram", "tiktok"]:
            improvements.extend([
                f"{base_desc} Hurry, limited stock! 🔥 Swipe up to get yours!",
                f"{base_desc} 😍 Join the trend - link in bio! ✨",
                f"{base_desc} Going viral! Get yours before they sell out! 🚀"
            ])
        elif channel.lower() == "linkedin":
            improvements.extend([
                f"{base_desc} Drive measurable business growth and ROI.",
                f"{base_desc} Trusted by 500+ enterprise clients worldwide.",
                f"{base_desc} Schedule a demo to see the impact on your business."
            ])
        else:
            improvements.extend([
                f"{base_desc} Buy now and save big with free shipping!",
                f"{base_desc} Explore our full collection - satisfaction guaranteed!",
                f"{base_desc} Don't wait - limited time offer ending soon!"
            ])

        return improvements[:2]

    @staticmethod
    def generate_cta_improvements(cta: str, channel: str) -> List[str]:
        """Generate CTA improvement suggestions"""
        if channel.lower() == "google":
            return ["Shop Now & Save", "Buy Today - Free Shipping", "Get Best Price", "Order Now"]
        elif channel.lower() in ["instagram", "tiktok"]:
            return ["Swipe Up 🔥", "Get Yours! 😍", "Shop Now ✨", "Link in Bio 👆"]
        elif channel.lower() == "linkedin":
            return ["Schedule Demo", "Get Enterprise Quote", "Learn More", "Contact Sales"]
        else:
            return ["Shop Now", "Buy Now & Save", "Get Yours Today", "Order Now"]

    @staticmethod
    async def _optimize_campaign_budget_fallback(request: MLCampaignOptimizationRequest) -> MLCampaignOptimizationResponse:
        """Fallback optimization using rule-based logic"""
        logger.info("Using fallback campaign optimization logic")

        # Simple rule-based budget allocation
        channels = ["instagram", "google", "tiktok", "facebook", "youtube", "linkedin"]

        # Base allocation weights based on general performance
        weights = {
            "google": 0.3,      # High intent traffic
            "facebook": 0.25,   # Good reach and targeting
            "instagram": 0.2,   # Visual products
            "youtube": 0.15,    # Video content
            "linkedin": 0.05,   # B2B
            "tiktok": 0.05      # Younger demographics
        }

        # Adjust weights based on target demographics
        if request.age < 30:
            weights["tiktok"] += 0.1
            weights["instagram"] += 0.1
            weights["facebook"] -= 0.1
            weights["google"] -= 0.1

        # Apply budget allocation
        recommended_split = {}
        for channel in channels:
            recommended_split[channel] = round(request.total_budget * weights.get(channel, 0), 2)

        # Simple revenue prediction (2.5x multiplier)
        predicted_revenue = request.total_budget * 2.5
        predicted_roi = 1.5  # 150% ROI

        return MLCampaignOptimizationResponse(
            recommended_split=recommended_split,
            predicted_revenue=predicted_revenue,
            predicted_roi=predicted_roi,
            confidence_score=0.6,  # Lower confidence for fallback
            warning="Using rule-based optimization (Campaign ML model not available)"
        )

    @staticmethod
    async def _score_creative_content_fallback(request: MLCreativeScoreRequest) -> MLCreativeScoreResponse:
        """Fallback creative scoring using simple rules"""
        logger.info("Using fallback creative scoring logic")

        # Simple rule-based scoring
        title_score = min(10, max(1, len(request.title.split()) * 1.5))
        desc_score = min(10, max(1, len(request.description.split()) * 0.5))
        cta_score = 8.0 if any(word in request.cta.lower() for word in ["buy", "shop", "get", "try"]) else 5.0

        return MLCreativeScoreResponse(
            channel=request.channel,
            scores={
                "title": title_score,
                "description": desc_score,
                "cta": cta_score,
                "channel_fit": 7.0,
                "final": round((title_score + desc_score + cta_score + 7.0) / 4, 1)
            },
            feedback=["Using rule-based scoring (DistilBERT model not available)"],
            improvements={
                "title": [f"{request.title} - Limited Time Offer!"],
                "description": [f"{request.description} Shop now and save!"],
                "cta": ["Shop Now", "Get Yours Today", "Buy Now"]
            }
        )

    @staticmethod
    async def health_check() -> Dict[str, Any]:
        """Check ML service health"""
        return {
            "ml_dependencies_available": ML_AVAILABLE,
            "campaign_model_loaded": campaign_model is not None,
            "feature_columns_loaded": feature_columns is not None,
            "distilbert_model_loaded": distilbert_model is not None,
            "nlp_embedder_loaded": nlp_models.get('embedder') is not None,
            "nlp_paraphraser_loaded": nlp_models.get('paraphraser') is not None,
            "models_loaded": (campaign_model is not None and distilbert_model is not None),
            "fallback_mode": not ML_AVAILABLE or campaign_model is None or distilbert_model is None
        }

File: src/services/ml-service.ts
import { Campaign, OptimizationSuggestion, Creative, CreativeScore } from '@/lib/types';

// ML Service configuration
const ML_SERVICE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

interface CampaignOptimizationRequest {
  total_budget: number;
  aov: number;
  age: number;
  gender: string;
  income_level: string;
  creative_quality: number;
  campaign_days: number;
  target_margin: number;
}

interface CampaignOptimizationResponse {
  recommended_split: Record<string, number>;
  predicted_revenue: number;
  predicted_roi: number;
  confidence_score: number;
  warning?: string;
}

interface CreativeScoreRequest {
  channel: string;
  title: string;
  description: string;
  cta: string;
}

interface CreativeScoreResponse {
  channel: string;
  scores: {
    title: number;
    description: number;
    cta: number;
    channel_fit: number;
    final: number;
  };
  feedback: string[];
  improvements: Record<string, string[]>;
}

class MLServiceError extends Error {
  constructor(message: string, public statusCode?: number) {
    super(message);
    this.name = 'MLServiceError';
  }
}

async function callMLService<T>(endpoint: string, data?: any): Promise<T> {
  try {
    const response = await fetch(`${ML_SERVICE_URL}${endpoint}`, {
      method: data ? 'POST' : 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
      body: data ? JSON.stringify(data) : undefined,
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new MLServiceError(
        `ML Service error: ${response.status} ${errorText}`,
        response.status
      );
    }

    return await response.json();
  } catch (error) {
    if (error instanceof MLServiceError) {
      throw error;
    }
    throw new MLServiceError(`Failed to connect to ML service: ${error}`);
  }
}

export async function optimizeCampaignBudget(campaign: Campaign): Promise<{
  optimizedBudget: Record<string, number>;
  predictions: {
    revenue: number;
    roi: number;
    confidence: number;
  };
  suggestions: OptimizationSuggestion[];
}> {
  // Map campaign data to ML service format
  const avgAge = (campaign.targeting.ageRange.min + campaign.targeting.ageRange.max) / 2;
  
  // Map gender values
  const genderMap: Record<string, string> = {
    'male': 'man',
    'female': 'woman',
    'all': 'all'
  };
  
  // Map income values
  const incomeMap: Record<string, string> = {
    'low': 'low',
    'medium': 'high', // Map medium to high for ML model
    'high': 'high',
    'all': 'high' // Default to high
  };

  // Estimate creative quality based on campaign creatives
  let creativeQuality = 0.7; // Default
  if (campaign.creatives && campaign.creatives.length > 0) {
    const avgScore = campaign.creatives.reduce((sum, creative) => {
      return sum + (creative.score?.overall || 70);
    }, 0) / campaign.creatives.length;
    creativeQuality = avgScore / 100;
  }

  const request: CampaignOptimizationRequest = {
    total_budget: campaign.budget.total,
    aov: campaign.product.price,
    age: avgAge,
    gender: genderMap[campaign.targeting.gender] || 'all',
    income_level: incomeMap[campaign.targeting.income] || 'high',
    creative_quality: creativeQuality,
    campaign_days: campaign.budget.duration,
    target_margin: campaign.product.targetMargin / 100, // Convert percentage to decimal
  };

  const response = await callMLService<CampaignOptimizationResponse>('/api/ml/campaign/optimize', request);

  // Convert response to application format
  const channelMapping: Record<string, string> = {
    'instagram': 'instagram',
    'google': 'google-ads',
    'tiktok': 'tiktok',
    'facebook': 'facebook',
    'youtube': 'youtube',
    'linkedin': 'linkedin'
  };

  const optimizedBudget: Record<string, number> = {};
  for (const [mlChannel, amount] of Object.entries(response.recommended_split)) {
    const appChannel = channelMapping[mlChannel];
    if (appChannel) {
      optimizedBudget[appChannel] = amount;
    }
  }

  // Generate optimization suggestions based on ML results
  const suggestions: OptimizationSuggestion[] = [];
  
  // Compare current vs optimal allocation
  const currentAllocation = campaign.budget.channels;
  for (const [channel, optimalAmount] of Object.entries(optimizedBudget)) {
    const currentAmount = currentAllocation[channel as keyof typeof currentAllocation] || 0;
    const difference = optimalAmount - currentAmount;
    
    if (Math.abs(difference) > campaign.budget.total * 0.05) { // 5% threshold
      suggestions.push({
        type: 'budget_reallocation',
        title: difference > 0 ? `Increase ${channel} budget` : `Decrease ${channel} budget`,
        description: difference > 0 
          ? `Increase budget allocation to ${channel} by $${Math.abs(difference).toFixed(2)} for better ROI`
          : `Reduce budget allocation from ${channel} by $${Math.abs(difference).toFixed(2)} and reallocate`,
        impact: {
          roi_increase: response.predicted_roi * 100 * (Math.abs(difference) / campaign.budget.total),
          reach_increase: 10 * (Math.abs(difference) / campaign.budget.total),
          conversion_increase: 8 * (Math.abs(difference) / campaign.budget.total),
        },
        changes: {
          to_channel: difference > 0 ? channel as any : undefined,
          from_channel: difference < 0 ? channel as any : undefined,
          amount: Math.abs(difference),
        },
      });
    }
  }

  return {
    optimizedBudget,
    predictions: {
      revenue: response.predicted_revenue,
      roi: response.predicted_roi,
      confidence: response.confidence_score,
    },
    suggestions,
  };
}

export async function scoreCreativeWithML(creative: Creative): Promise<CreativeScore> {
  // Map channel names for ML service
  const channelMapping: Record<string, string> = {
    'instagram': 'instagram',
    'google-ads': 'google',
    'tiktok': 'tiktok',
    'facebook': 'facebook',
    'youtube': 'youtube',
    'linkedin': 'linkedin',
    'twitter': 'twitter',
    'email': 'email'
  };

  const request: CreativeScoreRequest = {
    channel: channelMapping[creative.channel] || creative.channel,
    title: creative.title,
    description: creative.description,
    cta: creative.callToAction,
  };

  const response = await callMLService<CreativeScoreResponse>('/api/ml/creative/score', request);

  // Convert ML response to application format
  const score: CreativeScore = {
    overall: response.scores.final,
    breakdown: {
      clarity: response.scores.title,
      urgency: response.scores.description,
      relevance: response.scores.channel_fit,
      callToAction: response.scores.cta,
    },
    suggestions: [
      ...response.feedback,
      ...(response.improvements.title || []),
      ...(response.improvements.description || []),
      ...(response.improvements.cta || []),
    ],
  };

  return score;
}

export async function generateCreativeImprovements(
  channel: string,
  title: string,
  description: string,
  cta: string
): Promise<{
  titleSuggestions: string[];
  descriptionSuggestions: string[];
  ctaSuggestions: string[];
  overallScore: number;
}> {
  const channelMapping: Record<string, string> = {
    'instagram': 'instagram',
    'google-ads': 'google',
    'tiktok': 'tiktok',
    'facebook': 'facebook',
    'youtube': 'youtube',
    'linkedin': 'linkedin',
    'twitter': 'twitter',
    'email': 'email'
  };

  const request: CreativeScoreRequest = {
    channel: channelMapping[channel] || channel,
    title,
    description,
    cta,
  };

  const response = await callMLService<CreativeScoreResponse>('/api/ml/creative/score', request);

  return {
    titleSuggestions: response.improvements.title || [],
    descriptionSuggestions: response.improvements.description || [],
    ctaSuggestions: response.improvements.cta || [],
    overallScore: response.scores.final,
  };
}

// Health check for ML service
export async function checkMLServiceHealth(): Promise<boolean> {
  try {
    const response = await callMLService<{ status: string }>('/api/ml/health');
    return response.status === 'healthy' || response.status === 'degraded';
  } catch (error) {
    console.warn('ML Service health check failed:', error);
    return false;
  }
}

// Fallback functions when ML service is unavailable
export async function optimizeCampaignBudgetFallback(campaign: Campaign): Promise<{
  optimizedBudget: Record<string, number>;
  predictions: { revenue: number; roi: number; confidence: number };
  suggestions: OptimizationSuggestion[];
}> {
  // Simple rule-based optimization as fallback
  const totalBudget = campaign.budget.total;
  const numChannels = campaign.channels.preferred.length || 6;
  const equalAllocation = totalBudget / numChannels;

  const optimizedBudget: Record<string, number> = {};
  campaign.channels.preferred.forEach(channel => {
    optimizedBudget[channel] = equalAllocation;
  });

  // If no preferred channels, use default allocation
  if (campaign.channels.preferred.length === 0) {
    optimizedBudget['google-ads'] = totalBudget * 0.3;
    optimizedBudget['facebook'] = totalBudget * 0.25;
    optimizedBudget['instagram'] = totalBudget * 0.25;
    optimizedBudget['youtube'] = totalBudget * 0.2;
  }

  return {
    optimizedBudget,
    predictions: {
      revenue: totalBudget * 2.5, // Simple 2.5x revenue estimate
      roi: 1.5, // 150% ROI estimate
      confidence: 0.6, // Lower confidence for fallback
    },
    suggestions: [{
      type: 'creative_improvement',
      title: 'Improve creative quality',
      description: 'Consider A/B testing your ad creatives to improve performance',
      impact: {
        roi_increase: 15,
        reach_increase: 10,
        conversion_increase: 12,
      },
      changes: {
        creative_changes: ['Test different headlines', 'Improve call-to-action']
      },
    }],
  };
}